cirq-google/cirq_google/engine/stream_manager.py

```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
+            parent=project_name,
+            create_quantum_job=quantum.CreateQuantumJobRequest(
+                parent=program.name, quantum_job=job
+            ),
+        )
+        get_result_request = quantum.QuantumRunStreamRequest(
+            parent=project_name, get_quantum_result=quantum.GetQuantumResultRequest(parent=job.name)
+        )
+
+        current_request = create_program_and_job_request
+        while True:
+            try:
+                current_request.message_id = self._generate_message_id()
+                response_future = self._response_demux.subscribe(current_request.message_id)
+                await self._request_queue.put(current_request)
+                response = await response_future
+
+            # Broken stream
+            except google_exceptions.GoogleAPICallError as e:
+                if _is_retryable_error(e):
+                    # Retry
+                    self._response_demux.unsubscribe(current_request.message_id)
+                    current_request = get_result_request
+                    continue
+                    # TODO(#5996) add exponential backoff
+                raise e
+
+            # Either when this request is canceled or the _manage_stream() loop is canceled.
+            except asyncio.CancelledError:
+                if response_future is not None:
+                    response_future.cancel()
+                    await self._cancel(job.name)
+                raise
+
+            # Response handling
+            if 'result' in response:
+                return response.result
+            elif 'job' in response:
+                return response.job
+            elif 'error' in response:
+                current_request = _decide_retry_request_or_raise(
+                    response.error,
+                    current_request,
+                    create_program_and_job_request,
+                    create_job_request,
+                )
+                continue
+            else:
+                raise ValueError(
+                    'The Quantum Engine response type is not recognized by this client. '
+                    'This may be due to an outdated version of cirq-google'
+                )
+
+    async def _cancel(self, job_name: str) -> None:
+        await self._grpc_client.cancel_quantum_job(quantum.CancelQuantumJobRequest(name=job_name))
+
+    def _generate_message_id(self) -> str:
+        message_id = str(self._next_available_message_id)
+        self._next_available_message_id += 1
+        return message_id
+
+
+def _decide_retry_request_or_raise(
+    error: quantum.StreamError,
+    current_request,
+    create_program_and_job_request,
+    create_job_request: quantum.QuantumRunStreamRequest,
+):
+    """Decide whether the given stream error is retryable.
+
+    If it is, returns the next stream request to send upon retry. Otherwise, raises an error.
+    """
+    if error.code == Code.PROGRAM_DOES_NOT_EXIST:
```

Comment:
Do we have Python 3.10 for match/case statements? https://docs.python.org/3/whatsnew/3.10.html#pep-634-structural-pattern-matching

Reply:
Good to know this feature is in Python now! Unfortunately Cirq still supports 3.9.